### PR TITLE
MODOAIPMH-292: Incorrect cursor in resumptionToken

### PR DESCRIPTION
**Pull request by IlliaDaliek**
Jira: https://issues.folio.org/browse/MODOAIPMH-292

### PURPOSE
Fix incorrect cursor value returned within resumption token. 
Add an index to the request_id column in order to improve the request processing with date range parameters. Previously without index, the request responded with the timeout error.

The autoincremented id field was added to the instances table and is used for ordering instead of using the insatnce_id column when selecting instance ids from the table. Performing ordering by insatnce_id is unstable by the reason that while we requesting batches the new instances are saved in the background and the latest saved ids may come up at the top of the SELECT query result which brokes the behavior of using resumption token. In simple words, the stable order using the insatnce_id column for the "ORDER BY" statement can be achieved only when we have all instances already saved. Therefore while we ordering them by instance_id at the time when they are saved the order is changed as well. The autoincremented id fixed this and guarantee that the order will not be broken while the new incoming instances are saved.

As well since the SRS may respond with 502 when we request 500 records, the re-requesting of srs calls were implemented. In total there are 50 attempts to re-request SRS and only when all 50 attempts are failed the oai-pmh request processing is going to be canceled.

### CODE REVIEW

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java [hunk 1]
```diff
@@ -162,7 +163,7 @@ private MarcWithHoldingsRequestHelper() {
       }
 
       updateRequestMetadataFuture.onSuccess(res -> {
-        boolean isFirstBatch = resumptionToken == null || request.getRequestId() == null;
```
Comment (IlliaDaliek) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:165]: The second condition is unnecessary here. The first one tells the same. Request object can contain both at once or no one of them.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java [hunk 2]
```diff
@@ -200,7 +201,7 @@ private void processBatch(Request request, Context context, Promise<Response> oa
           return;
         }
 
-        if (!instances.isEmpty() && instances.get(0).getString(INSTANCE_ID_FIELD_NAME).equals(request.getNextRecordId())) {
+        if (!firstBatch && (CollectionUtils.isNotEmpty(instances) && !instances.get(0).getString(INSTANCE_ID_FIELD_NAME).equals(request.getNextRecordId()))) {
```
Comment (IlliaDaliek) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:204]: We should respond with an error in the case when instance ids are NOT equal.

**Comment by craigmcnally** [comment]
It would be great if you could add an "Approach" section to the PR description, with summary list of what's being done here... 

e.g.
* Add an ID column to the instance_ids table for purposes of sorting / maintaining the same order on the initial and subsequent requests.  This is simple integer incrementing from 0, 1, 2, ... N.
* Fix the stale resumption token check such that .... etc.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java [hunk 3]
```diff
@@ -113,11 +115,11 @@
   public static final MarcWithHoldingsRequestHelper INSTANCE = new MarcWithHoldingsRequestHelper();
   private final Vertx vertx;
 
-  public static final int POLLING_TIME_INTERVAL = 500;
-
-  public static final int MAX_WAIT_UNTIL_TIMEOUT = 1000*60*20;
+  private static final int REQUEST_ATTEMPTS = 50;
```
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:118]: Can you explain why these were changed to private?

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java (hunk 2 above)
Comment (craigmcnally) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:205]: Consider adding a comment in the code instead of here in the PR.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java [hunk 4]
```diff
@@ -219,12 +222,8 @@ private void processBatch(Request request, Context context, Promise<Response> oa
         final SourceStorageSourceRecordsClient srsClient = new SourceStorageSourceRecordsClient(request.getOkapiUrl(),
           request.getTenant(), request.getOkapiToken());
 
-        Future<Map<String, JsonObject>> srsResponse = Future.future();
-        if (CollectionUtils.isNotEmpty(instances)) {
-          srsResponse = requestSRSByIdentifiers(srsClient, instancesWithoutLast, deletedRecordSupport);
-        } else {
-          srsResponse.complete();
-        }
+        Future<Map<String, JsonObject>> srsResponse = requestSRSByIdentifiers(srsClient, instancesWithoutLast, deletedRecordSupport);
```
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:225]: probably don't need this `srsResponse` object at all, just call `onSuccess(...)` directly on the future returned by `requestSRSByIdentifiers(...)`

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java [hunk 5]
```diff
@@ -677,45 +674,71 @@ private HttpClientRequest createEnrichInventoryClientRequest(HttpClient httpClie
...
-              .filter(Objects::nonNull)
-              .map(JsonObject.class::cast)
-              .forEach(jo -> result.put(jo.getJsonObject("externalIdsHolder")
-                .getString(INSTANCE_ID_FIELD_NAME), jo));
+      srsClient.postSourceStorageSourceRecords("INSTANCE", deletedRecordSupport, listOfIds, srsResponse -> {
+        int statusCode = srsResponse.statusCode();
+        String statusMessage = srsResponse.statusMessage();
+        if (statusCode != 200) {
+          if(statusCode >= 500) {
+            String warnMessage = "Got error response form SRS, status code: " + statusCode + ", status message: {}" + statusMessage;
```
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:690]:
```suggestion
            String warnMessage = "Got error response form SRS, status code: " + statusCode + ", status message: " + statusMessage;
```

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java [hunk 6]
```diff
@@ -677,45 +674,71 @@ private HttpClientRequest createEnrichInventoryClientRequest(HttpClient httpClie
...
+            logger.warn(warnMessage);
+            if (attemptsCount.get() <= 0) {
+              String errorMessage = "SRS didn't respond with expected status code after " + REQUEST_ATTEMPTS + " attempts. Canceling further request processing.";
+              srsClient.close();
+              handleException(promise, new IllegalStateException(errorMessage));
+              return;
+            }
+            logger.warn("Trying to request SRS again, attempts left: {}", attemptsCount.decrementAndGet());
+            doPostRequestToSrs(srsClient, deletedRecordSupport, listOfIds, attemptsCount, promise);
```
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:699]: We probably want to have some retry backoff instead of just immediately trying again.  Ideally it would be configurable, but to reduce scope creep we can use a static constant.  It should be at least a second or two IMO

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java (hunk 3 above)
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:118]: It is simple class fields, they should be private. My guess here is that these fields were created among the code and then they were moved to constant by IDEA hotkeys with selected default access as public.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java (hunk 4 above)
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:225]: Yes, you are right. Thank you for your note.

File: src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java (hunk 6 above)
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java]: Thank you for the suggestion. I added vertx timer here with a 1-second delay.

**Review by craigmcnally — APPROVED** [review]
